[JSC] Yarr ParenContext should be allocated from Stack

Constellation opened the pull request:
#### a721886ce56b2e7bd28d37d0a06c7a2dea0ea429
<pre>
[JSC] Yarr ParenContext should be allocated from Stack
<a href="https://bugs.webkit.org/show_bug.cgi?id=303297">https://bugs.webkit.org/show_bug.cgi?id=303297</a>
<a href="https://rdar.apple.com/165603668">rdar://165603668</a>

Reviewed by Dan Hecht.

This patch makes ParenContext allocation done from the stack instead of
limited sized buffer. This makes code simpler, and more efficient
because of large size of the stack. Furthermore,

1. Previously we were always setting up the entire buffer for freelist
   of ParenContext. This setup time took significant amount of time.
   This patch changes how we allocate ParenContext: we first allocate
   them from the stack and freelist is used after we free some of these
   contexts. This avoids unnecessary setup time.
2. We avoid using stackPointerRegister to access to the allocated
   CallFrame for RegExp. Instead, we are using callFrameRegister, which
   allows us to extend stackPointerRegister easily for (1).
3. We clean up prologue and epilogue by using calleeSaveRegisters set.
4. Use TrustedImm32 for surrogateTagMask etc. in ARM64 too. This is not
   worth having a register.
5. non ARM64 uses InvalidGPRReg for freelistRegister. And using
   matchingContext&apos;s field instead.

* JSTests/stress/regexp-paren-context-stack-exhaustion.js: Added.
(createDeeplyNestedPattern):
(test):
* Source/JavaScriptCore/runtime/RegExpInlines.h:
(JSC::RegExp::matchInline):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::acquireRegExpPatternContexBuffer): Deleted.
(JSC::VM::releaseRegExpPatternContexBuffer): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::tryReadUnicodeCharImpl):
(JSC::Yarr::tryReadUnicodeCharSlowImpl):
* Source/JavaScriptCore/yarr/YarrJIT.h:
* Source/JavaScriptCore/yarr/YarrJITRegisters.h:
* Source/JavaScriptCore/yarr/YarrMatchingContextHolder.h:
(JSC::Yarr::MatchingContextHolder::offsetOfFreeList):
(JSC::Yarr::MatchingContextHolder::stackLimit const):
(JSC::Yarr::MatchingContextHolder::freeList const):
(JSC::Yarr::MatchingContextHolder::MatchingContextHolder):
(JSC::Yarr::MatchingContextHolder::~MatchingContextHolder):
(JSC::Yarr::MatchingContextHolder::offsetOfPatternContextBuffer): Deleted.
(JSC::Yarr::MatchingContextHolder::offsetOfPatternContextBufferSize): Deleted.

Canonical link: <a href="https://commits.webkit.org/303725@main">https://commits.webkit.org/303725@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d622f485cc66e859b0d7c7d202dc644fcb6298a2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133446 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5947 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141002 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/40b36958-7013-4b5b-9d2f-d56dab9058da) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6461 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5812 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102082 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/906c4dca-812e-4e42-908d-a6c077270856) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136393 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/119620 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82878 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b9af8c88-97e9-4b04-bdba-69205c76c7fd) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125518 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/113565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/37729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143648 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/131957 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5617 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38317 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/5699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4816 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110640 "Passed tests") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115879 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20635 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/5672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164922 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/5518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69124 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/5761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/5628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -623,57 +613,43 @@ class YarrGenerator final : public YarrJITInfo {
-        m_hitMatchLimit.append(m_jit.branchTestPtr(MacroAssembler::Zero, m_regs.remainingMatchCount));
-        m_jit.move(m_regs.freelistRegister, result);
-        m_jit.loadPtr(MacroAssembler::Address(m_regs.freelistRegister, ParenContext::nextOffset()), m_regs.freelistRegister);
+        m_abortExecution.append(m_jit.branchPtr(MacroAssembler::Above, MacroAssembler::Address(m_regs.matchingContext, MatchingContextHolder::offsetOfStackLimit()), result));
```
Do we have a test case that executes this taken branch?

Constellation replied:
Added

dhecht commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -1223,52 +1199,52 @@ class YarrGenerator final : public YarrJITInfo {
         callFrameSize *= sizeof(void*);
-        if (callFrameSize / sizeof(void*) != m_pattern.m_body->m_callFrameSize)
+        if (callFrameSize / sizeof(void*) != originalCallFrameSize)
             CRASH();
```
Is there a reason we don't use RELEASE_ASSERT here?

Constellation replied:
We should use RELEASE_ASSERT. Changed.

dhecht commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -5064,64 +5036,58 @@ class YarrGenerator final : public YarrJITInfo {
+            if (mayCall() || m_containsNestedSubpatterns)
+                m_jit.emitFunctionPrologue();
+            else
+                m_jit.tagReturnAddress();
```
do tests exercise the useJITCage case?

Constellation replied:
Yes, all iOS code uses this path.

dhecht commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -5344,9 +5286,8 @@ class YarrGenerator final : public YarrJITInfo {
 
             // The argument will be in an offset that depends on the arch and the number of registers we pushed into the stack
```
this comment seems outdated

Constellation replied:
Removed.

dhecht commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -5513,16 +5466,11 @@ class YarrGenerator final : public YarrJITInfo {
-#if ENABLE(YARR_JIT_ALL_PARENS_EXPRESSIONS)
-        if (m_containsNestedSubpatterns)
-            m_jit.move(MacroAssembler::TrustedImm32(matchLimit), m_regs.remainingMatchCount);
-#endif
```
why is this no longer needed?

Constellation replied:
This is because this `inlined` case never have `m_containsNestedSubpatterns` true. We have a release assert `RELEASE_ASSERT(!m_containsNestedSubpatterns)` in this function, so this is dead code.

dhecht reviewed (approved):
r=me with comments/questions

MenloDorian commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -623,57 +613,43 @@ class YarrGenerator final : public YarrJITInfo {
+
+        // Try to allocate from freelist first.
+        MacroAssembler::Jump allocateFromStack;
+        if (m_regs.freelistRegister != InvalidGPRReg) {
```
Does this need to be a runtime check?  It seems the port will either have a `freelistRegister` or not.  Unfortunately, it looks like `m_regs` is a runtime variable.  Can this be a `constexpr` check instead?

Constellation replied:
Unfortunately, this cannot be constexpr since this register can be random one when this register set is not `YarrJITDefaultRegisters` (it is used for inlined RegExp code). While non `YarrJITDefaultRegisters` is used only when this code path is not taken, but we cannot make sure that this is true statically. So this is a runtime check.

Constellation committed:
[JSC] Yarr ParenContext should be allocated from Stack
https://bugs.webkit.org/show_bug.cgi?id=303297
rdar://165603668

Reviewed by Dan Hecht.

This patch makes ParenContext allocation done from the stack instead of
limited sized buffer. This makes code simpler, and more efficient
because of large size of the stack. Furthermore,

1. Previously we were always setting up the entire buffer for freelist
   of ParenContext. This setup time took significant amount of time.
   This patch changes how we allocate ParenContext: we first allocate
   them from the stack and freelist is used after we free some of these
   contexts. This avoids unnecessary setup time.
2. We avoid using stackPointerRegister to access to the allocated
   CallFrame for RegExp. Instead, we are using callFrameRegister, which
   allows us to extend stackPointerRegister easily for (1).
3. We clean up prologue and epilogue by using calleeSaveRegisters set.
4. Use TrustedImm32 for surrogateTagMask etc. in ARM64 too. This is not
   worth having a register.
5. non ARM64 uses InvalidGPRReg for freelistRegister. And using
   matchingContext's field instead.

* JSTests/stress/regexp-paren-context-stack-exhaustion.js: Added.
(createDeeplyNestedPattern):
(test):
* Source/JavaScriptCore/runtime/RegExpInlines.h:
(JSC::RegExp::matchInline):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::acquireRegExpPatternContexBuffer): Deleted.
(JSC::VM::releaseRegExpPatternContexBuffer): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::tryReadUnicodeCharImpl):
(JSC::Yarr::tryReadUnicodeCharSlowImpl):
* Source/JavaScriptCore/yarr/YarrJIT.h:
* Source/JavaScriptCore/yarr/YarrJITRegisters.h:
* Source/JavaScriptCore/yarr/YarrMatchingContextHolder.h:
(JSC::Yarr::MatchingContextHolder::offsetOfFreeList):
(JSC::Yarr::MatchingContextHolder::stackLimit const):
(JSC::Yarr::MatchingContextHolder::freeList const):
(JSC::Yarr::MatchingContextHolder::MatchingContextHolder):
(JSC::Yarr::MatchingContextHolder::~MatchingContextHolder):
(JSC::Yarr::MatchingContextHolder::offsetOfPatternContextBuffer): Deleted.
(JSC::Yarr::MatchingContextHolder::offsetOfPatternContextBufferSize): Deleted.

Canonical link: https://commits.webkit.org/303725@main
